File: factorymethod/myfactorymethod.cpp
#include <iostream>
using namespace std;

class Fruit {
    public:
        Fruit() {}
        virtual ~Fruit() {}
        virtual string Consume() const = 0;
};

class Banana : public Fruit {
    public:
        string Consume() const override {
            return "Peel and Eat.";
        }
};

class Apple : public Fruit {
    public:
        string Consume() const override {
            return "Cut and Eat.";
        }
};

class Consumer {
    public:
        Consumer() {}
        virtual ~Consumer() {}
        virtual Fruit *GetFruit() const = 0;
        string Consume() const {
            Fruit *fruit = this->GetFruit();
            string result = "Method of consumption " + fruit->Consume();
            delete fruit;

            return result;
        }
};

class BananaConsumer: public Consumer {
    public:
        Fruit *GetFruit() const override {
            return new Banana();
        }
};

class AppleConsumer: public Consumer {
    public:
        Fruit *GetFruit() const override {
            return new Apple();
        }
};


// Utility method, it is common to all consumers
// specific consumer is passed in
void Usage(const Consumer &consumer) {
    cout << consumer.Consume() << endl;
}

int main() {
    Consumer *bananaconsumer = new BananaConsumer();
    Usage(*bananaconsumer);     // option 1
    bananaconsumer->Consume();  // option 2

    Consumer *appleconsumer = new AppleConsumer();
    Usage(*appleconsumer);      // option 1
    appleconsumer->Consume();   // option 2
}

File: factorymethod/swift/myfactorymethod.swift
protocol Fruit {
    func consume() -> String
}

struct Apple: Fruit {
    func consume() -> String {
        return "Cut and Eat."
    }
}

struct Banana: Fruit {
    func consume() -> String {
        return "Peel and Eat."
    }
}

protocol Consumer {
    func consume() -> String
    func getFruit() -> Fruit
}

struct AppleConsumer: Consumer {
    func getFruit() -> Fruit {
        let apple = Apple()
        return apple
    }

    func consume() -> String {
        let f = getFruit()
        return f.consume()
    }
}

struct BananaConsumer: Consumer {
    func getFruit() -> Fruit {
        let apple = Banana()
        return apple
    }

    func consume() -> String {
        let f = getFruit()
        return f.consume()
    }
}

let appleconsumer = AppleConsumer()
print(appleconsumer.consume())

let bananaconsumer = BananaConsumer()
print(bananaconsumer.consume())